File: js/ds/BinaryTree.js
class BinaryTreeNode {
  constructor(value) {
    this.value = value ? value : 0;
    this.left = null;
    this.right = null;
  }
}

class BinaryTree {
  constructor() {
    this.root = null;
  }

  add(value) {
    if (!this.root) {
      this.root = new BinaryTreeNode(value);
      return;
    }

    let current = this.root;
    let newNode = new BinaryTreeNode(value);
    while (current) {
      if (value < current.value) {
        if (!current.left) {
          current.left = newNode;
          return;
        }
        current = current.left;
      }

      if (value > current.value) {
        if (!current.right) {
          current.right = newNode;
          return;
        }
        current = current.right;
      }
    }
  }

  find(value) {
    if (this.root.value === value) return true;
    let current = this.root;
    while (current) {
      if (current.value === value) return true;
      if (value > current.value) {
        current = current.right;
      } else {
        current = current.left;
      }
    }
    return false;
  }

  inOrder(root) {
    if (root) {
      this.inOrder(root.left);
      console.log(root.value);
      this.inOrder(root.right);
    }
    return;
  }

  printInOrder() {
    let current = this.root;
    this.inOrder(current);
  }

  preOrder(root) {
    if (root) {
      console.log(root.value);
      this.inOrder(root.left);
      this.inOrder(root.right);
    }
  }

  printPreOrder() {
    let current = this.root;
    this.preOrder(current);
  }

  postOrder(root) {
    if (root) {
      this.postOrder(root.left);
      this.postOrder(root.right);
      console.log(root.value);
    }
  }

  printPostOrder() {
    let current = this.root;
    this.postOrder(current);
  }
}

const bt = new BinaryTree();
bt.add(5);
bt.add(6);
bt.add(2);
bt.add(1);
bt.add(3);
bt.add(4);
bt.add(8);
bt.add(9);
// console.log(bt);

// console.log(bt.find(3));
// bt.printInOrder();
// bt.printPreOrder();
bt.printPostOrder();

File: python/practice/ds/binary_tree.py
import random


class Node:
    def __init__(self, value=None, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right


class BinaryTree:
    def __init__(self, root=None):
        self.root = root

    def add(self, value):
        # if binary tree has no root node
        if self.root is None:
            new_node = Node(value)
            self.root = new_node
            return

        current = self.root

        while current:
            if value <= current.value:
                if current.left is None:
                    current.left = Node(value)
                    return
                else:
                    current = current.left

            if value >= current.value:
                if current.right is None:
                    current.right = Node(value)
                    print(current.right)
                    return
                else:
                    current = current.right

    def print_in_order(self):
        current = self.root
        self._in_order(current)

    def _in_order(self, node):
        if node is not None:
            self._in_order(node.left)
            print(node.value)
            self._in_order(node.right)

    def print_pre_order(self):
        current = self.root
        self._pre_order(current)

    def _pre_order(self, node):
        if node is None:
            return
        if node is not None:
            print(node.value)
            self._pre_order(node.left)
            self._pre_order(node.right)

    def print_post_order(self):
        current = self.root
        self._post_order(current)

    def _post_order(self, node):
        if node:
            self._post_order(node.left)
            self._post_order(node.right)
            print(node.value)


bt = BinaryTree()
bt.add(5)
bt.add(6)
bt.add(2)
bt.add(1)
bt.add(3)
bt.add(4)


bt.print_in_order()
print('======')
bt.print_pre_order()
print('======')
bt.print_post_order()